zephyr/test/vmh.c

```diff
@@ -17,8 +18,258 @@
+		alloc_size = min_alloc_size +
+		k_cycle_get_32() % (max_alloc_size - min_alloc_size + 1);
+
+		ptrs[i] = vmh_alloc(heap, alloc_size);
+
+		zassert_true(ptrs[i] != NULL,
+		"Allocation of size %u expected to succeed but failed",
+		alloc_size);
+
+	if (ptrs[i] != NULL) {
+		LOG_INF("Test allocation size: %d", alloc_size);
+
+		memset(ptrs[i], 0xAA, alloc_size);
+
+
+		uint8_t *byte_ptr = (uint8_t *)ptrs[i];
+
+		for (uint32_t j = 0; j < alloc_size; j++) {
+			zassert_equal(byte_ptr[j], 0xAA,
+			"Memory content verification failed at position %u", j);
+		}
+	}
+	}
+
+	for (int i = 0; i < num_allocs; i++) {
+		if (ptrs[i]) {
+			ret = vmh_free(heap, ptrs[i]);
+			zassert_equal(ret, 0, "Failed to free allocated memory");
+		}
+	}
+}
+
+/* Test case for multiple allocations */
+static void test_vmh_alloc_multiple_times(bool allocating_continuously)
+{
+	struct vmh_heap *heap =
+	vmh_init_heap(NULL, MEM_REG_ATTR_CORE_HEAP, 0, allocating_continuously);
+
+	zassert_not_null(heap, "Heap initialization failed");
+
+	/* Test multiple allocations with small sizes */
+	test_vmh_multiple_allocs(heap, 64, 4, 8);
+	test_vmh_multiple_allocs(heap, 64, 4, 8);
+	test_vmh_multiple_allocs(heap, 64, 4, 8);
+	test_vmh_multiple_allocs(heap, 16, 4, 1024);
+	test_vmh_multiple_allocs(heap, 16, 4, 1024);
+	test_vmh_multiple_allocs(heap, 64, 1024, 4096);
+	test_vmh_multiple_allocs(heap, 64, 4096, 8192);
+	test_vmh_multiple_allocs(heap, 16, 1024, 4096);
+	test_vmh_multiple_allocs(heap, 16, 4096, 8192);
+
+	/* Clean up the heap after testing */
+	int ret = vmh_free_heap(heap);
+
+	zassert_equal(ret, 0, "Failed to free heap after multiple allocations");
+}
+
+/* Test case for vmh_alloc and vmh_free */
+static void test_vmh_alloc_free(bool allocating_continuously)
+{
+	struct vmh_heap *heap =
+	vmh_init_heap(NULL, MEM_REG_ATTR_CORE_HEAP, 0, allocating_continuously);
+
+	zassert_not_null(heap, "Heap initialization failed");
+
+	test_vmh_alloc_free_no_check(heap, 512, true);
+	test_vmh_alloc_free_no_check(heap, 2048, true);
+	test_vmh_alloc_free_no_check(heap, sizeof(int), false);
+	test_vmh_alloc_free_no_check(heap, 0, false);
+
+	test_vmh_alloc_free_check(heap, 512, true);
+	test_vmh_alloc_free_check(heap, 2048, true);
+	test_vmh_alloc_free_check(heap, sizeof(int), false);
+	test_vmh_alloc_free_check(heap, 0, false);
+
+	int ret = vmh_free_heap(heap);
+
+	zassert_equal(ret, 0, "Failed to free heap");
+
+	/* Could add tests with configs for heaps*/
+}
+
+/* Test case for vmh_alloc and vmh_free with and without config */
+static void test_heap_creation(void)
+{
+	test_vmh_init_and_free_heap(MEM_REG_ATTR_CORE_HEAP, NULL, 0, false, true);
+
+	/* Try to setup with pre defined heap config */
+	struct vmh_heap_config config;
+
+	config.block_bundles_table[0].block_size = 4;
+
+	config.block_bundles_table[0].number_of_blocks = 80;
+
+	config.block_bundles_table[1].block_size = 4;
+
+	config.block_bundles_table[1].number_of_blocks = 80;
```

Comment:
if you're already testing 2 bundles, maybe configure them differently too?

Reply:
Also good catch on the bug in tests ;) block size can only be power of 2.